zabbix_auto_config/processing.py

```diff
@@ -205,20 +205,53 @@ def work(self) -> None:
         try:
             self.collect()
         except Exception as e:
-            logging.error("Collect exception: %s", str(e))
-            self.error_counter.add(exception=e)
-            if self.error_counter.tolerance_exceeded():
-                if self.config.exit_on_error:
-                    logging.critical(
-                        "Error tolerance exceeded. Terminating application."
-                    )
-                    self.stop_event.set()
-                    # TODO: raise exception with message above or just an empty exception?
-                else:
-                    self.disable()
-            raise ZACException(
-                f"Failed to collect from source {self.name!r}: {e}"
-            ) from e
+            self.handle_error(e)
+        else:
+            self.handle_success()
+
+    def increase_update_interval(self) -> None:
+        """Increase the update interval by the backoff factor."""
+        self.update_interval *= self.config.backoff_factor
+        logging.info(
+            "%s: Backing off. Increased update interval to %.2f",
+            self.name,
+            self.update_interval,
+        )
+
+    def reset_update_interval(self) -> None:
+        """Reset the update interval to its original value."""
+        if self.update_interval == self.config.update_interval:
+            return  # Nothing to do
+        self.update_interval = self.config.update_interval
+        logging.info(
+            "%s: Reset update interval to %.2f",
+            self.name,
+            self.update_interval,
+        )
+
+    def handle_success(self) -> None:
+        """Handle a successful collection."""
+        self.reset_update_interval()
+
+    def handle_error(self, e: Exception) -> None:
+        """Handle exceptions raised during collection."""
+        logging.error("Collect exception: %s", str(e))
+        self.error_counter.add(exception=e)
+
+        if self.error_counter.tolerance_exceeded():
```

Comment:
This is a bit flaky potentially. I would ideally like to decouple the error counter from strategies that do not require an error counter.

Reply:
Something like:

```py
strat = self.config.failure_strategy

if strat == models.FailureStrategy.DISABLE:
    # do stuff with error counter here
    ....
else:
    if strat == models.FailureStrategy.BACKOFF:
        self.increase_update_interval()
    # etc ...
```